src/Microsoft.ML.Transforms/Text/CharTokenizeTransform.cs

```diff
@@ -151,266 +140,319 @@ public override void Save(ModelSaveContext ctx)
+             : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
             {
-                // Create key values corresponding to the character. This will
-                // often just be the character itself, but sometimes (control characters,
-                // illegal codepoints, spaces, etc.) it is better to use something else
-                // to represent the character.
-                int[] boundaries = new int[CharsCount + 1];
-                var bldr = new StringBuilder();
-                for (int i = 1; i <= CharsCount; i++)
+                _parent = parent;
+                _types = new ColumnType[_parent.ColumnPairs.Length];
+                _isSourceVector = new bool[_parent.ColumnPairs.Length];
+                var keyType = new KeyType(DataKind.U2, 1, CharsCount);
+                for (int i = 0; i < _types.Length; i++)
                 {
-                    AppendCharRepr((char)i, bldr);
-                    boundaries[i] = bldr.Length;
+                    _isSourceVector[i] = inputSchema[_parent.ColumnPairs[i].input].Type.IsVector;
+                    _types[i] = new VectorType(keyType);
                 }
-
-                Host.Assert(bldr.Length == boundaries[boundaries.Length - 1]);
-                Interlocked.CompareExchange(ref _keyValuesBoundaries, boundaries, null);
-                Interlocked.CompareExchange(ref _keyValuesStr, bldr.ToString(), null);
-                bldr.Length = 0;
             }
 
-            var keyValuesStr = _keyValuesStr;
-            var keyValuesBoundaries = _keyValuesBoundaries;
-            Host.AssertValue(keyValuesBoundaries);
-
-            var values = dst.Values;
-            if (Utils.Size(values) < CharsCount)
-                values = new ReadOnlyMemory<char>[CharsCount];
-            for (int i = 0; i < CharsCount; i++)
-                values[i] = keyValuesStr.AsMemory().Slice(keyValuesBoundaries[i], keyValuesBoundaries[i + 1] - keyValuesBoundaries[i]);
-            dst = new VBuffer<ReadOnlyMemory<char>>(CharsCount, values, dst.Indices);
-        }
-
-        private void AppendCharRepr(char c, StringBuilder bldr)
-        {
-            // Special handling of characters identified in https://en.wikipedia.org/wiki/Unicode_control_characters,
-            // as well as space, using the control pictures.
-            if (c <= 0x20)
+            public override Schema.Column[] GetOutputColumns()
             {
-                // Use the control pictures unicode code block.
-                bldr.Append('<');
-                bldr.Append((char)(c + '\u2400'));
-                bldr.Append('>');
-                return;
+                var result = new Schema.Column[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    InputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out int colIndex);
```

Comment:
>colIndex [](start = 88, length = 8)

why do you need `colIndex`?

Reply:
BTW the assert below is incorrect. `TryGetColumnIndex` doesn't actually set `colIndex` to negative if it returns false, it sets it to 0.

---
In reply to: [226766006](https://github.com/dotnet/machinelearning/pull/1294#discussion_r226766006) [](ancestors = 226766006)